ambry-tools/src/main/java/com.github.ambry/clustermap/HelixBootstrapUpgradeTool.java

```diff
@@ -0,0 +1,470 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Properties;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import joptsimple.ArgumentAcceptingOptionSpec;
+import joptsimple.OptionParser;
+import joptsimple.OptionSet;
+import joptsimple.OptionSpec;
+import org.apache.helix.HelixAdmin;
+import org.apache.helix.manager.zk.ZKHelixAdmin;
+import org.apache.helix.model.IdealState;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.helix.model.LeaderStandbySMD;
+import org.apache.helix.model.builder.AutoModeISBuilder;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+
+/**
+ * This tool takes the hardware layout, partition layout and the Zk hosts information json files as input,
+ * and updates the ZK hosts with the contents of the layout files. It adds all partitions and hosts that have not
+ * previously been added (so, initially this will bootstrap the cluster information and on an ongoing basis, this can
+ * add new nodes and partitions).
+ *
+ * The existing hardware and partition layout json files will be read in as is.
+ *
+ * The ZkLayoutPath argument containing the Zk hosts information in each datacenter should be a json of the
+ * following example form:
+ *
+ * {
+ *   "zkHosts" : [
+ *     {
+ *       "datacenter":"dc1",
+ *       "hostname":"abc.example.com",
+ *       "port":"2199",
+ *     },
+ *     {
+ *       "datacenter":"dc2",
+ *       "hostname":"def.example.com",
+ *       "port":"2300",
+ *     }
+ *   ]
+ * }
+ *
+ * This tool should be run from an admin node that has access to the nodes in the hardware layout. The access is
+ * required because the static {@link ClusterMapManager} that we use to parse the static layout files validates
+ * these nodes.
+ *
+ * The tool does the following:
+ * 1. Bootstraps a static cluster map by adding nodes and partitions to Helix.
+ * 2. Upgrades the information with changes in the static clustermap that involve new nodes and new partitions.
+ *    To avoid over-complicating things, it assumes that the existing partition assignment does not change during an
+ *    upgrade. Newly added partitions can be distributed in any way (new partitions can have replicas even in
+ *    previously added nodes).
+ * 3. Upgrades will also update the partition states if required (READ_WRITE to SEALED or vice versa) for existing
+ *    partitions.
+ *
+ */
+public class HelixBootstrapUpgradeTool {
+  private final ClusterMapManager staticClusterMap;
+  private final Map<String, String> dataCenterToZkAddress = new HashMap<>();
+  private final Map<String, HelixAdmin> adminForDc = new HashMap<>();
+  // A mapping from partitions to Helix resource, populated at the start with what is present in Helix.
+  private final TreeMap<Long, Long> existingPartitions = new TreeMap<>();
```

Comment:
changed this to just a set of existing partitions, as the partition to resource mapping was unnecessary.
These are the partitions and resource that are already present in Helix when the tool starts up. That way I think "existing" is appropriate. These are not modified after initialization.